File: Problem28.java
//Time Complexity: O(n)
//Space Complexity: O(1)
// Did this code successfully run on Leetcode : Yes
// 238 Product of Array Except Self
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] output = new int[n];
        
        // Step 1: Calculate the left products
        output[0] = 1; // The product of elements to the left of index 0 is 1 (no elements)
        for (int i = 1; i < n; i++) {
            output[i] = output[i - 1] * nums[i - 1];
        }

        // Step 2: Calculate the right products and combine with left products
        int rightProduct = 1; // The product of elements to the right of index n-1 is 1 (no elements)
        for (int i = n - 1; i >= 0; i--) {
            output[i] *= rightProduct; // Combine left and right products
            rightProduct *= nums[i]; // Update the right product
        }

        return output;
    }
}

File: problem 28.py
# https://leetcode.com/problems/product-of-array-except-self/description/

# Time Complexity : O(n)
# Space Complexity : O(3n) ~ O(n)
# Did this code successfully run on Leetcode : YES

# Three line explanation of solution in plain english:
# took a prefix multiplication array and calculated the running multiplication in the left excluding itself
# and stored in it, similarly did it for swuffix as well
# and in the result we need to multiply the left and right from the corresponding arrays for the result


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        pm = [None] * n
        pm[0] = 1
        for i in range(1,n):
            pm[i] = pm[i-1] * nums[i-1]
        
        sm = [None] * (n)
        sm[n-1] = 1
        for i in range(n-2, -1, -1):
            sm[i] = sm[i+1] * nums[i+1]
        
        result = [None] * (n)
        for i in range(n):
            result[i] = pm[i] * sm[i]
        
        return result


# solution 2: with just one list (space optimization)

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [None] * n
        result[0]=1
        for i in range(1,n):
            result[i] = result[i-1] * nums[i-1]
        #print(result)
        
        mul = 1
        for i in range(n-2, -1, -1):
            mul *= nums[i+1]
            result[i] = result[i] * mul
        
        return result